File: benchmarks/tsb/bench_pct_change_periods.ts
/**
 * Benchmark: pctChangeSeries / pctChangeDataFrame with various period values.
 * Outputs JSON: {"function": "pct_change_periods", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, pctChangeSeries, pctChangeDataFrame } from "../../src/index.ts";

const ROWS = 100_000;
const WARMUP = 5;
const ITERATIONS = 30;

let s = 7;
const rand = () => {
  s = (s * 1664525 + 1013904223) & 0x7fffffff;
  return s / 0x7fffffff;
};

const data = Array.from({ length: ROWS }, () => rand() * 100 + 10);
const series = new Series({ data });

const df = new DataFrame({
  a: data,
  b: data.map((v) => v * 1.5),
  c: data.map((v) => v * 0.8),
});

for (let i = 0; i < WARMUP; i++) {
  pctChangeSeries(series, { periods: 1 });
  pctChangeSeries(series, { periods: 7 });
  pctChangeDataFrame(df, { periods: 5 });
}

const times: number[] = [];
for (let i = 0; i < ITERATIONS; i++) {
  const t0 = performance.now();
  pctChangeSeries(series, { periods: 1 });
  pctChangeSeries(series, { periods: 7 });
  pctChangeDataFrame(df, { periods: 5 });
  times.push(performance.now() - t0);
}

const total_ms = times.reduce((a, b) => a + b, 0);
console.log(
  JSON.stringify({
    function: "pct_change_periods",
    mean_ms: Math.round((total_ms / ITERATIONS) * 1000) / 1000,
    iterations: ITERATIONS,
    total_ms: Math.round(total_ms * 1000) / 1000,
  }),
);

File: benchmarks/pandas/bench_pct_change_periods.py
"""Benchmark: Series.pct_change() / DataFrame.pct_change() with various periods."""
import json, time
import pandas as pd
import numpy as np

ROWS = 100_000
WARMUP = 5
ITERATIONS = 30

rng = np.random.default_rng(7)
data = rng.random(ROWS) * 100 + 10

series = pd.Series(data)
df = pd.DataFrame({
    "a": data,
    "b": data * 1.5,
    "c": data * 0.8,
})

for _ in range(WARMUP):
    series.pct_change(periods=1)
    series.pct_change(periods=7)
    df.pct_change(periods=5)

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    series.pct_change(periods=1)
    series.pct_change(periods=7)
    df.pct_change(periods=5)
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({
    "function": "pct_change_periods",
    "mean_ms": round(total_ms / ITERATIONS, 3),
    "iterations": ITERATIONS,
    "total_ms": round(total_ms, 3),
}))
